test/validation/api/pktio/pktio.c

```diff
@@ -2709,6 +2709,36 @@ static void pktio_test_chksum_in_ipv4(void)
 			  pktio_test_chksum_in_ipv4_test);
 }
 
+static void pktio_test_chksum_bad_in_ipv4_prep(odp_packet_t pkt)
+{
+	odp_u16sum_t chksum;
+	uint32_t offset;
+
+	odph_ipv4_csum_update(pkt);
+
+	/* Insert incorrect checksum */
+	offset = odp_packet_l3_offset(pkt);
+	odp_packet_copy_to_mem(pkt,
+					offset + ODPH_IPV4HDR_CSUM_OFFSET,
+					2, &chksum);
+	chksum ^= 1;
+	odp_packet_copy_from_mem(pkt,
+					offset + ODPH_IPV4HDR_CSUM_OFFSET,
+					2, &chksum);
+}
+
+static void pktio_test_chksum_bad_in_ipv4_test(odp_packet_t pkt)
+{
+	CU_ASSERT(odp_packet_l3_chksum_status(pkt) == ODP_PACKET_CHKSUM_BAD);
+}
+
+static void pktio_test_chksum_bad_in_ipv4(void)
+{
+	pktio_test_chksum(pktio_test_chksum_in_ipv4_config,
```

Comment:
So pktio_test_chksum() function can't be used as-is for this test, because it does L4 checks, and these packets have an incorrect L3 checksum. It's a little more work to implement this one. It's up to you if you want to do it - if not, then please just remove this test.

Reply:
I think pktio_test_chksum is correct on its way for how it is now testing L3 checksum, if it is EXPECTED that, the implementation platform (linux_gen) should still return packet parser header with L4 offset set correctly and when the L3 proto is found to be a known protocol as well. It is just L3 header that got corrupted and we have L3_chksum_status API to validate this from application point of view. So, is the expectation on linux_gen correct as above ?. If so, we should fix this platform and hence I proposed fix to return right offset. Would you agree ?

Reply:
Even if L4 offset is set in case L3 checksum is bad, it's not clear (to me, at least) what the application or this validation test should or shouldn't do with it. In the general case, if L3 checksum is bad, then L4 offset could be wrong. In this particular test, we know that only the checksum is bad and the rest of ipv4 header is correct. But ODP implementation doesn't know that. Maybe @psavol can chime in here.

Reply:
API does not specify in which packet error cases implementation sets L4 offset with a correct value vs. offset_invalid. However, implementation should not set offset with a wrong value. Application should not assume that a packet has no errors, if L4 offset is set. It should use odp_packet_has_error(), etc instead. I.e. a packet may have errors (e.g. bad address) that does not prevent parser detecting where e.g. L4 starts.

Reply:
So, an additional assert for no other error  i,e odp_packet_has_error == 0 along with odp_packet_l4_chksum_status check in test side is sufficient I believe. I would not want any other portion of test frame work (like get_packets) to check for error as tests may intentionally test error packets in other places. Sound good ?

Reply:
@psavol @JereLeppanen ??.

Reply:
A test function for pktio_test_chksum() - such as pktio_test_chksum_bad_in_ipv4_test() could check that:
- chksum_status is ODP_PACKET_CHKSUM_BAD
- odp_packet_has_error() returns !=0
- odp_packet_has_l3_error() returns !=0 (an IP error)

But it should not check e.g. L4 offset, which it does not currently do.

Reply:
> So, an additional assert for no other error i,e odp_packet_has_error == 0 along with odp_packet_l4_chksum_status check in test side is sufficient I believe. I would not want any other portion of test frame work (like get_packets) to check for error as tests may intentionally test error packets in other places. Sound good ?

I meant odp_packet_l3_chksum_status here. Apologies.

Reply:
> A test function for pktio_test_chksum() - such as pktio_test_chksum_bad_in_ipv4_test() could check that:
> 
> * chksum_status is ODP_PACKET_CHKSUM_BAD
> * odp_packet_has_error() returns !=0
> * odp_packet_has_l3_error() returns !=0 (an IP error)
> 
> But it should not check e.g. L4 offset, which it does not currently do.

Right. I did not mean l4_chksum_status in ipv4 test and corrected my previous reply. Now, I am sending newer patch with above additional checks in test validation part. Could you please check. Thanks.

Reply:
@psavol I have added error and [l3/l4]_error checks in all checksum tests except in sctp, where l4_error() always == 0 meaning even if bad checksum is injected, this api never reported error. Up on checking, I think packet_l4_chksum() would have not set p.flags.l4_chksum_err correctly. I did not want to break CI for this known failure if I am correct and hence, skipped additional tests for sctp (bad checksum). Once I find root cause for it, i can send separate patch. Or we can tag it known failure (CU_PASS ??) in pktio_test_chksum_bad_in_sctp_test(), after adding l4_error check. What would you suggest ?. Thanks.

Reply:
Looks good to me, except that pktio_test_chksum_bad_in_ipv4_prep() could be done the same way as pktio_test_chksum_bad_in_udp_prep().

STCP fix and test case improvement can be done in another PR.

Reply:
Thanks @psavol and @JereLeppanen . Sent v8 addressing your valuable comments. Please check once.

Reply:
I think L4 offset is still being checked in the call chain pktio_test_chksum -> wait_for_packets -> wait_for_packets_hdr -> pktio_pkt_seq_hdr -> odp_packet_l4_offset. This does not cause a test failure because the offset is currently being set despite of incorrect L3 checksum. But as Petri stated, L4 checksum should not be checked in this case.

Speaking of which, since it's not required to set L4 offset when L3 checksum is bad, please revise the first commit in this PR accordingly, or simply drop the commit.

Reply:
True. wait_for_packets_hdr()/pktio_pkt_seq_hdr() needs to skip sequence number check when bad checksum is tested and there is a checksum error. The tricky part is that pktio_pkt_seq_hdr() is there to filter out packets originated from outside of the test case (e.g. on some pktios Linux IP stack sends periodic multicasts which this function filters out with sequence number checking).

Maybe some field in IP header could be used for filtering (in this bad checksum case), instead of magic number in UDP payload.

Reply:
I have modified the approach slightly to check return code (bad checksum can be ignored as the test injects them). Also, assumption is that, bad checksum can not be anything else bad than test is setting. Custom packet field may inject any side effect in case hw parser manipulates it (just a thought). Please see if this is fine.